<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Входные-данные:" data-toc-modified-id="Входные-данные:-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Входные данные:</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Разметка-каверов" data-toc-modified-id="Разметка-каверов-0.2.1"><span class="toc-item-num">0.2.1&nbsp;&nbsp;</span>Разметка каверов</a></span></li><li><span><a href="#Метаинформация" data-toc-modified-id="Метаинформация-0.2.2"><span class="toc-item-num">0.2.2&nbsp;&nbsp;</span>Метаинформация</a></span></li><li><span><a href="#Текст-песен" data-toc-modified-id="Текст-песен-0.2.3"><span class="toc-item-num">0.2.3&nbsp;&nbsp;</span>Текст песен</a></span></li></ul></li><li><span><a href="#Цель-проекта" data-toc-modified-id="Цель-проекта-0.3"><span class="toc-item-num">0.3&nbsp;&nbsp;</span>Цель проекта</a></span></li></ul></li><li><span><a href="#План-проекта" data-toc-modified-id="План-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>План проекта</a></span></li><li><span><a href="#EDA" data-toc-modified-id="EDA-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>EDA</a></span><ul class="toc-item"><li><span><a href="#Covers" data-toc-modified-id="Covers-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Covers</a></span></li><li><span><a href="#Lyrics" data-toc-modified-id="Lyrics-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Lyrics</a></span></li><li><span><a href="#Meta" data-toc-modified-id="Meta-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Meta</a></span></li><li><span><a href="#Объединённый-датафрейм" data-toc-modified-id="Объединённый-датафрейм-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Объединённый датафрейм</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка данных</a></span></li><li><span><a href="#Обучение-моделей" data-toc-modified-id="Обучение-моделей-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Обучение моделей</a></span><ul class="toc-item"><li><span><a href="#Baseline" data-toc-modified-id="Baseline-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Baseline</a></span></li><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Решающее дерево</a></span></li></ul></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Случайный лес</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#LGBM" data-toc-modified-id="LGBM-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>LGBM</a></span></li></ul></li><li><span><a href="#AdaBoost" data-toc-modified-id="AdaBoost-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>AdaBoost</a></span><ul class="toc-item"><li><span><a href="#Тестирование" data-toc-modified-id="Тестирование-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Тестирование</a></span></li></ul></li><li><span><a href="#Отчет-по-проекту" data-toc-modified-id="Отчет-по-проекту-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Отчет по проекту</a></span></li></ul></div>

# Хакатон. Обнаружение каверов музыкальных треков

### Входные данные:
Обнаружение треков каверов - важная продуктовая задача, которая может значительно улучшить качество рекомендаций музыкального сервиса и повысить счастье наших пользователей. Если мы умеем с высокой точностью классифицировать каверы и связывать их между собой, то можно предложить пользователю новые возможности для управления потоком треков. Например:

- по желанию пользователя можем полностью исключить каверы из рекомендаций;
- показать все каверы на любимый трек пользователя;
- контролировать долю каверов в ленте пользователя.


### Описание данных

#### Разметка каверов

Файл covers.json содержит разметку каверов, сделанную редакторами сервиса:

- track_id - уникальный идентификатор трека;
- track_remake_type - метка, присвоенная редакторами. Может принимать значения ORIGINAL и COVER;
- original_track_id - уникальный идентификатор исходного трека.

<aside>
💡 Обратите внимание, что не для всех каверов известны идентификаторы исходных треков!!!

</aside>

#### Метаинформация

- track_id - уникальный идентификатор трека;
- dttm - первая дата появления информации о треке;
- title - название трека;
- language - язык исполнения;
- isrc - международный уникальный идентификатор трека;
- genres - жанры;
- duration - длительность трека;

#### Текст песен

- track_id - уникальный идентификатор трека;
- lyricId - уникальный идентификатор текста;
- text - текст трека.

### Цель проекта

В рамках данного проекта необходимо провести классификацию музыкальных треков на кавер-версии и оригинальные композиции. Для решения этой задачи мы имеем свободу выбора сторонних библиотек и инструментов. В процессе исследования было принято решение использовать метрику f1 score.

Выбор f1 score обоснован тем, что она предоставляет комплексную оценку точности и полноты классификации. В контексте нашей задачи это крайне важно, поскольку неверно классифицированный трек может оставить негативное впечатление о качестве системы. Метрика f1 score позволяет учесть не только количество правильно или неправильно классифицированных треков, но и качество этих предсказаний.

## План проекта

Предстоит построить прототип модели машинного обучения. Его разработка состоит из нескольких этапов:

1. Исследовательский анализ данных.
- провести ознакомление с данными;
- по необходимости изменить тип данных;
- определить категориальные признаки;
- проверить данные на пропуски;
- заполнить пропуски по необходимости;
- выявить влияние категориальных признаков на целевой;
- выявить корреляцию между признаками;
- объединить данные и посмотреть на зависимости;
- сделать выводы по результатам анализа данных.

2. Подготовка данных для моделей.
- объединить все таблицы в одну;
- при необходимости добавить признаки;
- провести преобразование категориальных признаков в численные техникой OE;
- разделить данные на 2 выборки обучающую и тестовую в пропорции 3:1
- провести стандартизацию данных;
- избавиться от дисбалланса признаков
- провести оценку важности признаков
- провести анализ матрицы ошибок

3. Обучение моделей.
- Baseline;
- LogisticRegression;
- RandomForestClassifier;
- CatBoostClassifier
- применить стэкинг

4. Отчет по проведенной работе.

In [6]:
import os
import json
from concurrent.futures import ProcessPoolExecutor

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import optuna
from langdetect import detect
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, StackingClassifier
from sklearn.linear_model import LogisticRegression
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler, OrdinalEncoder
from sklearn.dummy import DummyClassifier
from sklearn.metrics import f1_score, make_scorer
from sklearn.feature_extraction.text import TfidfVectorizer
from scipy.sparse import csr_matrix
from sklearn.decomposition import PCA
from tqdm import tqdm

tqdm.pandas()

In [7]:
RANDOM_STATE = 42

Прочитаем предоставленные данные

In [8]:
def json_opening(path):
    data = []

    with open(path, encoding='utf-8') as datafile:
        for line in datafile:
            data.append(json.loads(line))

    return pd.DataFrame(data)


In [9]:
df_covers = json_opening('covers.json')
df_lyrics = json_opening('lyrics.json')
df_meta = json_opening('meta.json')

## EDA

### Covers

Проведем изучение датасета `Covers`

In [10]:
df_covers.head()

original_track_id                          track_id  \
0  eeb69a3cb92300456b6a5f4162093851  eeb69a3cb92300456b6a5f4162093851   
1  fe7ee8fc1959cc7214fa21c4840dff0a  fe7ee8fc1959cc7214fa21c4840dff0a   
2  cd89fef7ffdd490db800357f47722b20  cd89fef7ffdd490db800357f47722b20   
3  995665640dc319973d3173a74a03860c  995665640dc319973d3173a74a03860c   
4                              None  d6288499d0083cc34e60a077b7c4b3e1   

  track_remake_type  
0          ORIGINAL  
1          ORIGINAL  
2          ORIGINAL  
3          ORIGINAL  
4             COVER

In [11]:
df_covers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71597 entries, 0 to 71596
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   original_track_id  4821 non-null   object
 1   track_id           71597 non-null  object
 2   track_remake_type  71597 non-null  object
dtypes: object(3)
memory usage: 1.6+ MB


In [12]:
df_covers.describe()

original_track_id                          track_id  \
count                               4821                             71597   
unique                              4468                             71597   
top     f3e5e9db509f9b6f6d93b9c38e4e1324  2d09f68013fd6b3998f1163986877d2c   
freq                                  38                                 1   

       track_remake_type  
count              71597  
unique                 2  
top                COVER  
freq               67304

In [13]:
df_covers.isna().sum()

original_track_id    66776
track_id                 0
track_remake_type        0
dtype: int64

Все фичи имеют корректные названия. Огромное количество нанов в столбце `original_track_id`.

Всего треков в датасете сильно больше, чем оригинальных треков что закономерно.

Замечаем, что в столбце `original_track_id`	есть почти 400 дубликатов. Видимо это как раз те оригиналы, на которых есть больше чем 1 кавер. Это странно и интересно, ведь информация в этом столбце это не айди трека, которого кавер, а исключительно айди оригинальных треков. Чтобы разобраться, откуда дубликаты, сначала предположим, что `original_track_id` есть только у тех треков, которые с лейблом `ORIGINAL`. Посмотрим, есть ли такие id оригиналов, у которых есть метка `COVER`.

In [14]:
df_covers.groupby('track_remake_type')['original_track_id'].count()

track_remake_type
COVER        528
ORIGINAL    4293
Name: original_track_id, dtype: int64

Так и есть, получаем 528 треков в `original_track_id` с меткой `COVER`. Это даже больше, чем дубликатов в столбце. Оставим только эти треки и посмотрим, есть ли среди них дубликаты

In [15]:
duplicates = df_covers[df_covers['track_remake_type'] == 'COVER']['original_track_id'].duplicated(keep=False)
duplicate_tracks = df_covers[df_covers['track_remake_type'] == 'COVER'][duplicates]
duplicate_tracks['original_track_id'].count()

330

Получается 330 дубликатов. Это почти столько же, сколько Дубликатов по этому столбцу всего (353). Итак, теперь понятно, что всё-таки у нас есть несколько каверов, для которых мы уже знаем их id оригинального трека.

Ещё было бы не лишним посмотреть, действительно ли все оригинальные треки имеют свой `original_id`.

In [16]:
df_covers[df_covers['track_remake_type'] == 'ORIGINAL']['original_track_id'].isna().sum()

0

Да, так и есть. Все треки без `original_id` являются каверами.

In [17]:
df_covers.duplicated().sum()

0

При этом полные дубликаты в датасете тоже отсутствуют.

### Lyrics

In [18]:
df_lyrics.head()

lyricId  \
0  a951f9504e89759e9d23039b7b17ec14   
1  0c749bc3f01eb8e6cf986fa14ccfc585   
2  e2c8830fbc86e5964478243099eec23a   
3  e2c8830fbc86e5964478243099eec23a   
4  7624653ca8522ba93470843c74961b7d   

                                                text  \
0  Живу сейчас обломами, обломками не той любви\n...   
1  Tell me your fable\nA fable\nTell me your fabl...   
2  You're ashamed about all your fears and doubts...   
3  You're ashamed about all your fears and doubts...   
4  You showed him all the best of you,\nBut I'm a...   

                           track_id  
0  1c4b1230f937e4c548ff732523214dcd  
1  0faea89b0d7d6235b5b74def72511bd8  
2  9c6dc41d5ccd9968d07f055da5d8f741  
3  bfd04a73e9cffdf0e282c92219a86ea1  
4  8d70930d09cd239c948408d1317d8659

In [19]:
df_lyrics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11414 entries, 0 to 11413
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   lyricId   11414 non-null  object
 1   text      11414 non-null  object
 2   track_id  11414 non-null  object
dtypes: object(3)
memory usage: 267.6+ KB


In [20]:
df_lyrics.describe()

lyricId  \
count                              11414   
unique                             10915   
top     6fe05ce48761f833737487cd7a499cd3   
freq                                   9   

                                                     text  \
count                                               11414   
unique                                              10644   
top     Twenty-five years and my life is still\nI'm tr...   
freq                                                   15   

                                track_id  
count                              11414  
unique                             10277  
top     b891000882f4a612190142861b5c038b  
freq                                  24

In [21]:
df_lyrics.isna().sum()

lyricId     0
text        0
track_id    0
dtype: int64

Пропусков нет, все столбцы с правильным названием. Переименуем `lyricId` в `lyric_id`.

Замечаем, что присутствуют дубликаты в стобце `track_id`, который должен быть уникальным. В результате анализа обнаружили, что это связано с тем, что некоторые тексты песен различаются лишь несколькими символами, хотя по сути являются одинаковыми. Избавимся от этих дубликатов.

In [22]:
df_lyrics.drop_duplicates(subset='track_id', keep='first', inplace=True)

In [23]:
df_lyrics.rename(columns={'lyricId': 'lyric_id'}, inplace=True)

In [24]:
df_lyrics.describe()

lyric_id  \
count                              10277   
unique                              9891   
top     28fbafaf626c497d0d4d933b4f0d3f5b   
freq                                   9   

                                                     text  \
count                                               10277   
unique                                               9852   
top     I don't know just how it happened, I let down ...   
freq                                                    9   

                                track_id  
count                              10277  
unique                             10277  
top     7d98246460d93188a9cac6ba41c15138  
freq                                   1

Готово

### Meta

In [25]:
df_meta.head() 

track_id          dttm                 title  \
0  c3b9d6a354ca008aa4518329aaa21380  1.639688e+12        Happy New Year   
1  c57e3d13bbbf5322584a7e92e6f1f7ff  1.637762e+12            Bad Habits   
2  955f2aafe8717908c140bf122ba4172d  1.637768e+12  Por Esa Loca Vanidad   
3  fae5a077c9956045955dde02143bd8ff  1.637768e+12          Mil Lagrimas   
4  6bede082154d34fc18d9a6744bc95bf5  1.637768e+12   Sexo Humo y Alcohol   

  language          isrc             genres  duration  
0       EN  RUB422103970            [DANCE]  161120.0  
1       EN  QZN882178276      [ELECTRONICS]  362260.0  
2     None  QZNJZ2122549  [FOLK, LATINFOLK]  260000.0  
3     None  QZNJZ2166033  [FOLK, LATINFOLK]  190000.0  
4     None  QZNJZ2122551  [FOLK, LATINFOLK]  203000.0

In [26]:
df_meta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71769 entries, 0 to 71768
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   track_id  71768 non-null  object 
 1   dttm      71768 non-null  float64
 2   title     71768 non-null  object 
 3   language  21969 non-null  object 
 4   isrc      71455 non-null  object 
 5   genres    71768 non-null  object 
 6   duration  71768 non-null  float64
dtypes: float64(2), object(5)
memory usage: 3.8+ MB


In [27]:
df_meta.isna().sum()

track_id        1
dttm            1
title           1
language    49800
isrc          314
genres          1
duration        1
dtype: int64

In [28]:
df_meta.sort_values(by='track_id').tail()

track_id          dttm                   title  \
35829  fffc8f02feca70f7fad88fb987b76c4e  1.633135e+12  La Luz de la Cara Roja   
36591  fffefc046267a3f6263fc2c8d0f3f871  1.633989e+12          Uska Hi Banana   
71472  ffff8f1c01d78267ce067a2507fcc8f8  1.558112e+12                No Limit   
46002  ffff9a22aa441db772206280ffb8b52e  1.423084e+12          You're a Woman   
44434                              None           NaN                    None   

      language          isrc           genres  duration  
35829     None  ARAB82100052  [ROCK, ALLROCK]  200100.0  
36591     None  QZNJX2142230           [FOLK]  185880.0  
71472     None  PLS921752714               []  175950.0  
46002     None          None            [POP]  321820.0  
44434     None          None             None       NaN

Названия правильные, дата не в неправильном формате, приведём к `datetime`.

Присутсвуют пропуски в столбцах `language` (очень много) и `isrc`.

Есть одна полностью пустая строка, удалим ее.

Язык будем восстанавливать по названию трека.

Пропуски в `isrc` никак не восстановить - удалим.

In [29]:
df_meta.dropna(how='all', inplace=True)

In [30]:
df_meta['dttm'] = pd.to_datetime(df_meta['dttm'], unit='ms')

Теперь создадим функцию исключение, чтобы `langdetector` не выдал ошибку. Применяем функцию, заполняем пропуски языка по названию треков

In [31]:
%%time
def detect_language(text):
    try:
        result = detect(text)
        return result.upper()
    except:
        return None

Wall time: 0 ns


In [32]:
df_meta['language'].fillna(
    df_meta['title'].progress_apply(
        lambda x: detect_language(str(x)) if pd.notna(x) and x.strip() else None
    ),
    inplace=True
)

100%|████████████████████████████████████████████████████████████████████████████| 71768/71768 [22:04<00:00, 54.18it/s]


In [33]:
df_meta.isna().sum()

track_id      0
dttm          0
title         0
language     69
isrc        313
genres        0
duration      0
dtype: int64

Проверяем - готово, осталось всего 69 миссов! Отличный результат. Удаляем остатки

In [34]:
df_meta.dropna(inplace=True)

Проверим, корректно ли заполнился язык.

In [35]:
df_meta.head(30)

track_id                dttm  \
0   c3b9d6a354ca008aa4518329aaa21380 2021-12-16 21:00:00   
1   c57e3d13bbbf5322584a7e92e6f1f7ff 2021-11-24 14:00:33   
2   955f2aafe8717908c140bf122ba4172d 2021-11-24 15:37:03   
3   fae5a077c9956045955dde02143bd8ff 2021-11-24 15:37:03   
4   6bede082154d34fc18d9a6744bc95bf5 2021-11-24 15:37:03   
5   551331ee1444de6658e71933eb7f57a2 2021-11-24 15:37:03   
6   2be8bb01bb1aeef0908524dc11935a88 2021-11-24 15:37:03   
7   3883d1add3ff8b60e77548a64df7c892 2021-11-24 15:37:03   
8   59aa886979699aa7e4e2a8aed43bfe03 2021-11-24 15:37:03   
9   2a38992b7769963937246288ad89b44c 2021-11-24 15:37:03   
10  530c73cb44a3b3abd7f2b299977fbb19 2021-11-24 15:37:03   
11  c11320ee81872399e23a2670749d393a 2021-11-24 15:37:03   
12  5f95372a5ae6e6d0d7606ea11dce1ece 2021-11-24 15:37:03   
13  adfa176e2fb0b952569129a22d64de4d 2021-11-24 15:37:03   
14  a7309d9c3374e12201f3226bc0fa7980 2021-11-24 15:37:21   
15  1069db8b42f2f79b6a66ae57c6332670 2021-11-24 16:47:17   
16  f973c844c174b9eafca7fa87560c2df9 2021-11-24 16:47:18   
17  c996348ac49e0c243a4eb4cb0a9cfbbb 2021-11-24 16:57:06   
18  55a4890cc667212a048ce81b4049d1ea 2021-11-24 16:59:05   
19  f8de56ba1399dfab829176023352024d 2021-11-24 17:03:24   
20  c0d52a0b4713145bcd3b10b7e5315bfa 2021-11-24 17:03:24   
21  0cd778cff703e232d0626ee47ec8c0ae 2021-11-24 17:03:24   
22  0c0315f45a49a35901912d0084cdf03f 2021-11-24 17:03:24   
23  b07061f3d66aee20c96ed970a46c227b 2021-11-24 17:03:24   
24  6baf269773c2bff0aa080991de62115d 2021-11-24 17:03:24   
25  94093810e71074bec5a6990c1c214165 2022-10-19 21:00:00   
26  2b74d8dc05d365e51cf4ff2bcc9727f0 2021-12-09 21:00:00   
27  bdca41f86284a6284c8e48c440bcced0 2021-11-24 17:42:00   
28  988b3d43fc47fdf843b6c21de0fe3bb4 2021-11-24 17:59:32   
29  b83e2289c343f6196c08c2b48a9c689f 2021-11-24 17:59:40   

                                                title language          isrc  \
0                                      Happy New Year       EN  RUB422103970   
1                                          Bad Habits       EN  QZN882178276   
2                                Por Esa Loca Vanidad       PT  QZNJZ2122549   
3                                        Mil Lagrimas       IT  QZNJZ2166033   
4                                 Sexo Humo y Alcohol       ES  QZNJZ2122551   
5                                       La Ultima Vez       IT  QZNJZ2166034   
6                     Homenaje a Pastor de los Santos       ES  QZNJZ2122553   
7                                      Los Marginales       ES  QZNJZ2122554   
8                                      Cuenta Conmigo       ES  QZNJZ2122555   
9                                      Ahora Juego Yo       ES  QZNJZ2122556   
10                                           Mariposa       ES  QZNJZ2166035   
11  Voy a Beber / Piensas en Mi /  Por el Momento ...       ES  QZNJZ2166036   
12                                   Adicto a Tu Piel       EN  QZNJZ2166037   
13                                  Mi Amor Imposible       CA  QZNJZ2166038   
14                  Canto para Pedir Posada Cumbiamix       PT  QZNJZ2167682   
15                                         The Moment       EN  QZNJZ2173087   
16                                    Forever in Love       NO  QZNJZ2173055   
17                               Boa Noite Vizinhança       PT  QZNJZ2178383   
18                                      Tacones Rojos       ES  GBKPL2174553   
19                                  Parabéns pra Você       PT  QZNJZ2137727   
20                                            Timidez       RO  QZNJZ2137728   
21                                             Trovão       PT  QZNJZ2137729   
22                                 Inigualável Paixão       PT  QZNJZ2137730   
23                                 Pra Você Acreditar       RO  QZNJZ2137731   
24                                        Pé na Areia       PT  QZNJZ2137732   
25                                           Humsafar       HI  GX3HH2123738   
26     

Визуально всё в правильном формате, и почти всегда попадает в точку. Проверяем остатки пропусков.

In [36]:
df_meta.isna().sum()

track_id    0
dttm        0
title       0
language    0
isrc        0
genres      0
duration    0
dtype: int64

Готово! теперь посмотрим дубликаты

In [37]:
df_meta.drop(columns=['genres'], axis=1, inplace=False).duplicated().sum()

0

In [38]:
df_meta['track_id'].duplicated().sum()

0

In [39]:
df_meta['isrc'].duplicated().sum()

172

Полные дубликаты отсутствуют, `track_id` все уникальные. Присутсвуют дубликаты в `isrc`, который по идее должен быть уникальным. Скорее всего это разные релизы одного и того же трека. Удалим их.

In [40]:
df_meta.drop_duplicates(subset='isrc', keep='first', inplace=True)

In [41]:
df_meta['isrc'].duplicated().sum() #проверяем

0

Готово! Начальный анализ проведён.

**Вывод:** у нас есть три датасета, содержащих разную информацию. От пропусков избавились (кроме `original_track_id`), аномальные дубликаты удалены. Теперь можно сделать из них единый датафрейм. Уникальный `track_id` назначим индексом.

### Объединённый датафрейм

In [42]:
df = df_covers.merge(df_lyrics, on='track_id', how='outer').merge(df_meta, on='track_id', how='outer')

df.set_index('track_id', inplace=True)

In [43]:
df.head()

original_track_id  \
track_id                                                             
eeb69a3cb92300456b6a5f4162093851  eeb69a3cb92300456b6a5f4162093851   
fe7ee8fc1959cc7214fa21c4840dff0a  fe7ee8fc1959cc7214fa21c4840dff0a   
cd89fef7ffdd490db800357f47722b20  cd89fef7ffdd490db800357f47722b20   
995665640dc319973d3173a74a03860c  995665640dc319973d3173a74a03860c   
d6288499d0083cc34e60a077b7c4b3e1                              None   

                                 track_remake_type  \
track_id                                             
eeb69a3cb92300456b6a5f4162093851          ORIGINAL   
fe7ee8fc1959cc7214fa21c4840dff0a          ORIGINAL   
cd89fef7ffdd490db800357f47722b20          ORIGINAL   
995665640dc319973d3173a74a03860c          ORIGINAL   
d6288499d0083cc34e60a077b7c4b3e1             COVER   

                                                          lyric_id  \
track_id                                                             
eeb69a3cb92300456b6a5f4162093851  260f21d9f48e8de874a6e844159ddf28   
fe7ee8fc1959cc7214fa21c4840dff0a  2498827bd11eca5846270487e4960080   
cd89fef7ffdd490db800357f47722b20  5237001311d4062bf2b80de30652bf58   
995665640dc319973d3173a74a03860c  e5b1b57090b728e8d98d2b4d9b781bf4   
d6288499d0083cc34e60a077b7c4b3e1                               NaN   

                                                                               text  \
track_id                                                                              
eeb69a3cb92300456b6a5f4162093851  Left a good job in the city\nWorkin' for the m...   
fe7ee8fc1959cc7214fa21c4840dff0a  Some folks are born made to wave the flag\nOoh...   
cd89fef7ffdd490db800357f47722b20  Uno por pobre y feo, hombre\nPero antoja'o, ay...   
995665640dc319973d3173a74a03860c  Yeah!... yeah!... remember the time, baby... y...   
d6288499d0083cc34e60a077b7c4b3e1                                                NaN   

                                                dttm            title  \
track_id                                                                
eeb69a3cb92300456b6a5f4162093851 2009-11-11 21:00:00       Proud Mary   
fe7ee8fc1959cc7214fa21c4840dff0a 2009-11-11 21:00:00    Fortunate Son   
cd89fef7ffdd490db800357f47722b20 2009-09-21 20:00:00  La Camisa Negra   
995665640dc319973d3173a74a03860c 2009-11-16 21:00:00    The Way I Are   
d6288499d0083cc34e60a077b7c4b3e1 2009-09-17 20:00:00        Extraball   

                                 language          isrc  \
track_id                                                  
eeb69a3cb92300456b6a5f4162093851       EN  USFI86900049   
fe7ee8fc1959cc7214fa21c4840dff0a       EN  USFI86900065   
cd89fef7ffdd490db800357f47722b20       ES  USUL10400965   
995665640dc319973d3173a74a03860c       EN  USUM70722806   
d6288499d0083cc34e60a077b7c4b3e1       EN  FR8Q10900116   

                                                  genres  duration  
track_id                                                            
eeb69a3cb92300456b6a5f4162093851         [ROCK, ALLROCK]  187220.0  
fe7ee8fc1959cc7214fa21c4840dff0a         [ROCK, ALLROCK]  137780.0  
cd89fef7ffdd490db800357f47722b20  [POP, FOLK, LATINFOLK]  216840.0  
995665640dc319973d3173a74a03860c       [FOREIGNRAP, RAP]  179660.0  
d6288499d0083cc34e60a077b7c4b3e1           [ELECTRONICS]  212620.0

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 71767 entries, eeb69a3cb92300456b6a5f4162093851 to b0e056b8cdd428972b758e8ea8a0ce8c
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   original_track_id  4821 non-null   object        
 1   track_remake_type  71597 non-null  object        
 2   lyric_id           10277 non-null  object        
 3   text               10277 non-null  object        
 4   dttm               71215 non-null  datetime64[ns]
 5   title              71215 non-null  object        
 6   language           71215 non-null  object        
 7   isrc               71215 non-null  object        
 8   genres             71215 non-null  object        
 9   duration           71215 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(8)
memory usage: 6.0+ MB


In [45]:
df.isna().sum()

original_track_id    66946
track_remake_type      170
lyric_id             61490
text                 61490
dttm                   552
title                  552
language               552
isrc                   552
genres                 552
duration               552
dtype: int64

После объединения появились пропуски. Что касается `id` оригиналов и текста - ничего удивительного, а откуда в остальных колонках? Посмотрим. Выводим хвост датафрейма, отсортированного по названию (где точно не должно по идее быть пропусков)

In [46]:
df.sort_values(by='title').tail(20)

original_track_id  \
track_id                                                             
e03f31f1286208dcb0cf1e5ee84aceff  e03f31f1286208dcb0cf1e5ee84aceff   
c0308f925cc2c6d8b70aef91e7b8f338                              None   
23a6c682503e8a9829f70b5f8925c62b                              None   
8f9dd56bdeb9ada69cd291899f52d5a8  8f9dd56bdeb9ada69cd291899f52d5a8   
571a376663b9e86159b2234841700cda                              None   
36dc45aa77dc758427f3aaab066d569b                              None   
753a6567816f1f1fd458caeaeb33ea6a  753a6567816f1f1fd458caeaeb33ea6a   
eccece56e9409fbc59cfab0a5c0d1787  eccece56e9409fbc59cfab0a5c0d1787   
137f3f182530fb4ba015ef2ce69eedfc  137f3f182530fb4ba015ef2ce69eedfc   
ac71c0295c370a352f170431e1156b0f  ac71c0295c370a352f170431e1156b0f   
c63d43ce874ea17a1e98aa4ae526af73  c63d43ce874ea17a1e98aa4ae526af73   
7ee22bc05263ac26e24e5f895e1dbc72                              None   
e710cc10abdd5b16107c71d231a52df1  e710cc10abdd5b16107c71d231a52df1   
197694c04d7d6b2426cdebad6f06b57a  197694c04d7d6b2426cdebad6f06b57a   
341c141769af816e128d3906fd5b90f8  341c141769af816e128d3906fd5b90f8   
3d4b4c7adc6b80270711517280ad22e3                              None   
176a983bf850334e5c36d965b726a6e0                              None   
953ef412c3fd8e6698219e7e486af561                              None   
872140dcd8a61120041ef75f53188e76                               NaN   
f994d35681afefa37e8448c5e90a28e9                               NaN   

                                 track_remake_type  \
track_id                                             
e03f31f1286208dcb0cf1e5ee84aceff          ORIGINAL   
c0308f925cc2c6d8b70aef91e7b8f338             COVER   
23a6c682503e8a9829f70b5f8925c62b             COVER   
8f9dd56bdeb9ada69cd291899f52d5a8          ORIGINAL   
571a376663b9e86159b2234841700cda             COVER   
36dc45aa77dc758427f3aaab066d569b             COVER   
753a6567816f1f1fd458caeaeb33ea6a          ORIGINAL   
eccece56e9409fbc59cfab0a5c0d1787          ORIGINAL   
137f3f182530fb4ba015ef2ce69eedfc          ORIGINAL   
ac71c0295c370a352f170431e1156b0f          ORIGINAL   
c63d43ce874ea17a1e98aa4ae526af73          ORIGINAL   
7ee22bc05263ac26e24e5f895e1dbc72             COVER   
e710cc10abdd5b16107c71d231a52df1          ORIGINAL   
197694c04d7d6b2426cdebad6f06b57a          ORIGINAL   
341c141769af816e128d3906fd5b90f8          ORIGINAL   
3d4b4c7adc6b80270711517280ad22e3             COVER   
176a983bf850334e5c36d965b726a6e0             COVER   
953ef412c3fd8e6698219e7e486af561             COVER   
872140dcd8a61120041ef75f53188e76               NaN   
f994d35681afefa37e8448c5e90a28e9               NaN   

                                                          lyric_id  \
track_id                                                             
e03f31f1286208dcb0cf1e5ee84aceff                               NaN   
c0308f925cc2c6d8b70aef91e7b8f338                               NaN   
23a6c682503e8a9829f70b5f8925c62b                               NaN   
8f9dd56bdeb9ada69cd291899f52d5a8                               NaN   
571a376663b9e86159b2234841700cda                               NaN   
36dc45aa77dc758427f3aaab066d569b                               NaN   
753a6567816f1f1fd458caeaeb33ea6a  8a05c998880f8ae66cbd708c0537ed1b   
eccece56e9409fbc59cfab0a5c0d1787                               NaN   
137f3f182530fb4ba015ef2ce69eedfc                               NaN   
ac71c0295c370a352f170431e1156b0f                               NaN   
c63d43ce874ea17a1e98aa4ae526af73                               NaN   
7ee22bc05263ac26e24e5f895e1dbc72                               NaN   
e710cc10abdd5b16107c71d231a52df1  cd57f5fb44ec0887dc1b30cf44b0b8e9   
197694c04d7d6b2426cdebad6f06b57a  f431426dcfeec5f2dee9516ff3e85755   
341c141769af816e128d3906fd5b90f8                               NaN   
3d4b4c7adc6b80270711517280ad22e3                               NaN   
176a983bf850334e5c36d965b726a6e0                           

Видим, что в датафрейме присутствуют строки, про которые нет никакой информации, кроме `id` трека и метки оригинал\кавер. Можно было бы заполнить эту информацию по `id` оригинального трека, но он есть только у треков с меткой `ORIGINAL`, и скорее всего такой информации просто нет в датасете. К тому же, здесь присутствуют строки вовсе без `id` оригинала, только с меткой `COVER`. Придётся удалить эти строки из датасета. Теперь посмотрим как выглядят строки с пропусками по метке кавер\оригинал.

In [47]:
df.sort_values(by='track_remake_type').tail(20)

original_track_id track_remake_type  \
track_id                                                               
1e8ca836c962598551882e689265c1c5               NaN               NaN   
1296f487d135fa2808e18accebb6554f               NaN               NaN   
5a1fce1ebb19da07487f3132b261b78e               NaN               NaN   
a33e4b17ce2c355fb3e65ee042509803               NaN               NaN   
0e2b32a929c0e09569fdd0c831e2b459               NaN               NaN   
a6bea0f1418b5e768b689604e11d111d               NaN               NaN   
fb68073731d1fb23269c9778df343711               NaN               NaN   
96437a3517f1f408381363bc8d539718               NaN               NaN   
694ddd9c20051cd893222f00497b58a7               NaN               NaN   
ac8c3425ac530ce79214dd0704a08e30               NaN               NaN   
0f74b97c7afdd09412b1147e812aad76               NaN               NaN   
7684d25699831e0e09c923ec709f602c               NaN               NaN   
39282a5bda389deac70c6c33ebd05fe7               NaN               NaN   
105e505f39a1ce3a9e277e81783a8ffc               NaN               NaN   
43bcd59d680a05b9ff28993e6375cb64               NaN               NaN   
fa1b6f8b3ba1609dfaf6bda5b180ef25               NaN               NaN   
5e16b48a01483936ae404c3b30286a02               NaN               NaN   
b8d4dc8f7abdd59df56eb39787b35c8f               NaN               NaN   
b9233cb60674ca1ebb868e1cfe8d7b5c               NaN               NaN   
b0e056b8cdd428972b758e8ea8a0ce8c               NaN               NaN   

                                                          lyric_id  \
track_id                                                             
1e8ca836c962598551882e689265c1c5  bccf43f213b3abecaa0bdcdc17f5fff7   
1296f487d135fa2808e18accebb6554f  144248a3cb888ffc4f28c4b3df9f45c1   
5a1fce1ebb19da07487f3132b261b78e  6a3fdd4107a56ad1d9d622246b6ba327   
a33e4b17ce2c355fb3e65ee042509803  bc9d77c8771c3d81aa848adc339f1fb5   
0e2b32a929c0e09569fdd0c831e2b459                               NaN   
a6bea0f1418b5e768b689604e11d111d                               NaN   
fb68073731d1fb23269c9778df343711                               NaN   
96437a3517f1f408381363bc8d539718                               NaN   
694ddd9c20051cd893222f00497b58a7                               NaN   
ac8c3425ac530ce79214dd0704a08e30                               NaN   
0f74b97c7afdd09412b1147e812aad76                               NaN   
7684d25699831e0e09c923ec709f602c                               NaN   
39282a5bda389deac70c6c33ebd05fe7                               NaN   
105e505f39a1ce3a9e277e81783a8ffc                               NaN   
43bcd59d680a05b9ff28993e6375cb64                               NaN   
fa1b6f8b3ba1609dfaf6bda5b180ef25                               NaN   
5e16b48a01483936ae404c3b30286a02                               NaN   
b8d4dc8f7abdd59df56eb39787b35c8f                               NaN   
b9233cb60674ca1ebb868e1cfe8d7b5c                               NaN   
b0e056b8cdd428972b758e8ea8a0ce8c                               NaN   

                                                                               text  \
track_id                                                                              
1e8ca836c962598551882e689265c1c5  Off with your head\nDance til you're dead\nHea...   
1296f487d135fa2808e18accebb6554f  Your body, your soul\nFeel the rhythm\nTake co...   
5a1fce1ebb19da07487f3132b261b78e  Last night, all I think about is you\nDon't st...   
a33e4b17ce2c355fb3e65ee042509803  Нас не догонят\nНас не догонят, нас не догонят...   
0e2b32a929c0e09569fdd0c831e2b459                                                NaN   
a6bea0f1418b5e768b689604e11d111d                                                NaN   
fb68073731d1fb23269c9778df343711                                                NaN   
96437a3517f1f408381363bc8d539718                                                NaN   
694ddd9c20051cd893222f00497b58a7     

Уже лучше. Для всех треков без метки есть какая-то информация, навание, текст, жанры и тд. Но для всех выведенных строк также пропущено и значения оригинального трека. А так как ранее мы выяснили, что все треки без оригинального `id` являются каверами, можно заполнить пропуски в этом столбце значением `COVER`.

In [48]:
df['track_remake_type'].fillna('COVER', inplace=True)

In [49]:
df.dropna(subset=['title'], inplace=True) #Удаляем строки с пропущенной информацией о треках

In [50]:
df.isna().sum() #проверяем

original_track_id    66504
track_remake_type        0
lyric_id             61098
text                 61098
dttm                     0
title                    0
language                 0
isrc                     0
genres                   0
duration                 0
dtype: int64

В одном из выводов замечаем, что столбец `genres` содержит пустые списки. Это не пропуски, поэтому они могли ускользнуть от внимания и впоследствии создать проблемы при обучении моделей. Посмотрим, сколько всего таких значений в датасете.

In [51]:
len(df[df['genres'].apply(lambda x: len(x) == 0)])

2315

Довольно много, удалять нельзя. Преобразуем списки в строки, и заменим их заглушкой `Unknown`

In [52]:
df = df.explode('genres')

In [53]:
df['genres'].replace('', 'Unknown', inplace=True)

In [54]:
df1 = df.query("isrc == 'PLS921752714'") #Проверяем, как заполнились пустые списки
df1

original_track_id track_remake_type lyric_id  \
track_id                                                                        
ffff8f1c01d78267ce067a2507fcc8f8              None             COVER      NaN   

                                 text                dttm     title language  \
track_id                                                                       
ffff8f1c01d78267ce067a2507fcc8f8  NaN 2019-05-17 16:50:33  No Limit       SQ   

                                          isrc genres  duration  
track_id                                                         
ffff8f1c01d78267ce067a2507fcc8f8  PLS921752714    NaN  175950.0

In [55]:
df['text'].fillna('', inplace=True)

Исследуем баланс классов целевого признака

In [56]:
df['track_remake_type'].value_counts()

COVER       99758
ORIGINAL     6649
Name: track_remake_type, dtype: int64

Видим дисбаланс 16 к одному, что закономерно, так как к одному треку может быть много каверов. Но для модели это может создать проблемы, при обучении нужно будет определить сбалансированные веса.

## Предобработка данных

Для начала разобъем выборку на тренировочную и тестовую, чтобы избежать утечки данных.

In [57]:
df_train, df_test = train_test_split(df, test_size=0.25, random_state=RANDOM_STATE)

Создадим новые фичи на основе векторизации текста.

In [58]:
def vectorizer_text(df):
    # Параметры векторизации
    max_features = 200
    max_df = 0.80
    min_df = 3

    # Векторизация текста
    vectorizer = TfidfVectorizer(lowercase=True, max_features=max_features, max_df=max_df, min_df=min_df)
    matrix = vectorizer.fit_transform(df['text'])
    sparse_matrix = csr_matrix(matrix)

    # Создание позиций слов
    Positions = {}
    for j, word in enumerate(vectorizer.get_feature_names_out()):
        Positions[word] = j

    # Количество слов
    amountWords = len(vectorizer.get_feature_names_out())

    # Создание DataFrame из TF-IDF матрицы
    Tf_idf = pd.DataFrame.sparse.from_spmatrix(sparse_matrix, columns=vectorizer.get_feature_names_out())

    # Применение PCA для уменьшения размерности
    n_components = 30
    pca = PCA(n_components=n_components)
    matrix_pca = pca.fit_transform(sparse_matrix.toarray())
    
    
    pca_df = pd.DataFrame(matrix_pca, columns=[f'PCA_{i}' for i in range(matrix_pca.shape[1])])
    pca_df.index = df.index
    df = pd.concat([df, pca_df], axis=1)
    
    return df

In [59]:
df_train = vectorizer_text(df_train)
df_test = vectorizer_text(df_test)

In [60]:
df_train

original_track_id track_remake_type  \
track_id                                                               
458ce539d6ab93504964dfde4b9368dd              None             COVER   
a374d6de89e64007a1540d4f894a22a4              None             COVER   
15ae2f556bfacc3afaf77e6cf0eedb47              None             COVER   
c8ae5081580a4ca4699457ec0ceece28              None             COVER   
df375b046a3596832c71247ddfa6bc2a              None             COVER   
...                                            ...               ...   
51c4bc89bc21e9c06fa83af73cbd3512              None             COVER   
f9b887685587f0d2e220ddb88876e366              None             COVER   
2f4ec94bf2ff1ff9fbc6112ff6640a22              None             COVER   
60d2451205f75a41f1de12ed156b0273              None             COVER   
e59b698a6eca92186929890be10837e7              None             COVER   

                                                          lyric_id  \
track_id                                                             
458ce539d6ab93504964dfde4b9368dd  4fcc0b872f5c1fc8fbe7f8ec985f1c94   
a374d6de89e64007a1540d4f894a22a4                               NaN   
15ae2f556bfacc3afaf77e6cf0eedb47                               NaN   
c8ae5081580a4ca4699457ec0ceece28                               NaN   
df375b046a3596832c71247ddfa6bc2a                               NaN   
...                                                            ...   
51c4bc89bc21e9c06fa83af73cbd3512                               NaN   
f9b887685587f0d2e220ddb88876e366                               NaN   
2f4ec94bf2ff1ff9fbc6112ff6640a22                               NaN   
60d2451205f75a41f1de12ed156b0273  cb7c84bd5e0e50ad6d8f2cbbd9d4c8ea   
e59b698a6eca92186929890be10837e7                               NaN   

                                                                               text  \
track_id                                                                              
458ce539d6ab93504964dfde4b9368dd  Llévame donde los hombres\nNecesiten tus palab...   
a374d6de89e64007a1540d4f894a22a4                                                      
15ae2f556bfacc3afaf77e6cf0eedb47                                                      
c8ae5081580a4ca4699457ec0ceece28                                                      
df375b046a3596832c71247ddfa6bc2a                                                      
...                                                                             ...   
51c4bc89bc21e9c06fa83af73cbd3512                                                      
f9b887685587f0d2e220ddb88876e366                                                      
2f4ec94bf2ff1ff9fbc6112ff6640a22                                                      
60d2451205f75a41f1de12ed156b0273  We've already said goodbye\nSince you gotta go...   
e59b698a6eca92186929890be10837e7                                                      

                                                dttm             title  \
track_id                                                                 
458ce539d6ab93504964dfde4b9368dd 2020-12-06 00:24:30    Alma Misionera   
a374d6de89e64007a1540d4f894a22a4 2021-08-19 13:19:29       Slaughtered   
15ae2f556bfacc3afaf77e6cf0eedb47 2012-02-29 20:00:00    Getting Better   
c8ae5081580a4ca4699457ec0ceece28 2021-03-30 13:30:10               911   
df375b046a3596832c71247ddfa6bc2a 2014-06-11 02:37:25  Days in the West   
...                                              ...               ...   
51c4bc89bc21e9c06fa83af73cbd3512 2021-08-30 21:00:00    Pareja del Año   
f9b887685587f0d2e220ddb88876e366 2021-10-28 04:21:26      Like a Stone   
2f4ec94bf2ff1ff9fbc6112ff6640a22 2022-09-15 21:00:00        Bad Habits   
60d2451205f75a41f1de12ed156b0273 2014-09-20 20:00:00            Go Now   
e59b698a6eca92186929890be10837e7 2018-02-24 17:07:45        Heartbeats   

                                 language          isrc        genres  \
track_id         

In [61]:
df_test

original_track_id  \
track_id                                                             
89803636c59346850bfb6de14370b303  89803636c59346850bfb6de14370b303   
224cb225a2e612f5489b24992a20cbf3                              None   
12f22354aa53f7a47bac64322f8a8779                              None   
0f61b0f76abab917e60b82b7fb58c6d6                              None   
c996348ac49e0c243a4eb4cb0a9cfbbb                              None   
...                                                            ...   
5cb4a290db5564e663072da94d23ad17                              None   
73c65bc6ae2e2f09a02317c7bd25654d                              None   
19525276e1be80fd62e804864bcfe3cc                              None   
304eb9d73bd89118a0b598ed42194260                              None   
dbfd0061ef91d9267174288ad3044cca                              None   

                                 track_remake_type  \
track_id                                             
89803636c59346850bfb6de14370b303          ORIGINAL   
224cb225a2e612f5489b24992a20cbf3             COVER   
12f22354aa53f7a47bac64322f8a8779             COVER   
0f61b0f76abab917e60b82b7fb58c6d6             COVER   
c996348ac49e0c243a4eb4cb0a9cfbbb             COVER   
...                                            ...   
5cb4a290db5564e663072da94d23ad17             COVER   
73c65bc6ae2e2f09a02317c7bd25654d             COVER   
19525276e1be80fd62e804864bcfe3cc             COVER   
304eb9d73bd89118a0b598ed42194260             COVER   
dbfd0061ef91d9267174288ad3044cca             COVER   

                                                          lyric_id  \
track_id                                                             
89803636c59346850bfb6de14370b303  fed4f1741f5feebb79dd250921dcd02f   
224cb225a2e612f5489b24992a20cbf3                               NaN   
12f22354aa53f7a47bac64322f8a8779                               NaN   
0f61b0f76abab917e60b82b7fb58c6d6                               NaN   
c996348ac49e0c243a4eb4cb0a9cfbbb                               NaN   
...                                                            ...   
5cb4a290db5564e663072da94d23ad17                               NaN   
73c65bc6ae2e2f09a02317c7bd25654d                               NaN   
19525276e1be80fd62e804864bcfe3cc                               NaN   
304eb9d73bd89118a0b598ed42194260                               NaN   
dbfd0061ef91d9267174288ad3044cca                               NaN   

                                                                               text  \
track_id                                                                              
89803636c59346850bfb6de14370b303  Нам что покрепче, нас лечат бесконечные встреч...   
224cb225a2e612f5489b24992a20cbf3                                                      
12f22354aa53f7a47bac64322f8a8779                                                      
0f61b0f76abab917e60b82b7fb58c6d6                                                      
c996348ac49e0c243a4eb4cb0a9cfbbb                                                      
...                                                                             ...   
5cb4a290db5564e663072da94d23ad17                                                      
73c65bc6ae2e2f09a02317c7bd25654d                                                      
19525276e1be80fd62e804864bcfe3cc                                                      
304eb9d73bd89118a0b598ed42194260                                                      
dbfd0061ef91d9267174288ad3044cca                                                      

                                                dttm                   title  \
track_id                                                                       
89803636c59346850bfb6de14370b303 2022-06-23 21:00:00                 Счастье   
224cb225a2e612f5489b24992a20cbf3 2021-03-23 05:15:10            Hackearam-Me   
12f22354aa53f7a47bac64322f8a8779 2020-06-28 18:38:26  Daredevil - Main T

In [63]:
df_train.isna().sum()

track_id                 0
original_track_id    74283
track_remake_type        0
lyric_id             68090
text                     0
dttm                     0
title                    0
language                 0
isrc                     0
genres                1743
duration                 0
PCA_0                    0
PCA_1                    0
PCA_2                    0
PCA_3                    0
PCA_4                    0
PCA_5                    0
PCA_6                    0
PCA_7                    0
PCA_8                    0
PCA_9                    0
PCA_10                   0
PCA_11                   0
PCA_12                   0
PCA_13                   0
PCA_14                   0
PCA_15                   0
PCA_16                   0
PCA_17                   0
PCA_18                   0
PCA_19                   0
PCA_20                   0
PCA_21                   0
PCA_22                   0
PCA_23                   0
PCA_24                   0
PCA_25                   0
P

In [64]:
df_test.isna().sum()

track_id                 0
original_track_id    24783
track_remake_type        0
lyric_id             22760
text                     0
dttm                     0
title                    0
language                 0
isrc                     0
genres                 572
duration                 0
PCA_0                    0
PCA_1                    0
PCA_2                    0
PCA_3                    0
PCA_4                    0
PCA_5                    0
PCA_6                    0
PCA_7                    0
PCA_8                    0
PCA_9                    0
PCA_10                   0
PCA_11                   0
PCA_12                   0
PCA_13                   0
PCA_14                   0
PCA_15                   0
PCA_16                   0
PCA_17                   0
PCA_18                   0
PCA_19                   0
PCA_20                   0
PCA_21                   0
PCA_22                   0
PCA_23                   0
PCA_24                   0
PCA_25                   0
P

Удаляем лишние признаки

In [65]:
df_train = df_train.drop(['original_track_id', 'lyric_id', 'isrc'], axis=1)
df_test = df_test.drop(['original_track_id', 'lyric_id', 'isrc'], axis=1)

Создадим новый признак на основе даты (количество дней с момента релиза)

In [66]:
today = pd.to_datetime('2023-10-31')
df_train['days_since_release'] = (today - df_train['dttm']).dt.days
df_test['days_since_release'] = (today - df_test['dttm']).dt.days

Создадим новый признак на основе названия (количество символов в названии)

In [67]:
df_train['title'] = df_train['title'].str.len()
df_test['title'] = df_test['title'].str.len()

Создадим новый признак на основе текста (количество слов в тексте)

In [68]:
df_train['text'] = df_train['text'].str.split().str.len()
df_test['text'] = df_test['text'].str.split().str.len()

In [69]:
df_train = df_train.drop(['dttm'], axis=1)
df_test = df_test.drop(['dttm'], axis=1)

In [70]:
df_train.info()
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79805 entries, 0 to 79804
Data columns (total 38 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   track_id            79805 non-null  object 
 1   track_remake_type   79805 non-null  object 
 2   text                79805 non-null  int64  
 3   title               79805 non-null  int64  
 4   language            79805 non-null  object 
 5   genres              78062 non-null  object 
 6   duration            79805 non-null  float64
 7   PCA_0               79805 non-null  float64
 8   PCA_1               79805 non-null  float64
 9   PCA_2               79805 non-null  float64
 10  PCA_3               79805 non-null  float64
 11  PCA_4               79805 non-null  float64
 12  PCA_5               79805 non-null  float64
 13  PCA_6               79805 non-null  float64
 14  PCA_7               79805 non-null  float64
 15  PCA_8               79805 non-null  float64
 16  PCA_

In [71]:
df_train.head()

track_id track_remake_type  text  title language  \
0  458ce539d6ab93504964dfde4b9368dd             COVER   157     14       IT   
1  a374d6de89e64007a1540d4f894a22a4             COVER     0     11       EN   
2  15ae2f556bfacc3afaf77e6cf0eedb47             COVER     0     14       EN   
3  c8ae5081580a4ca4699457ec0ceece28             COVER     0      3       ES   
4  df375b046a3596832c71247ddfa6bc2a             COVER     0     16       EN   

       genres  duration     PCA_0     PCA_1     PCA_2  ...    PCA_21  \
0   SPIRITUAL  132590.0 -0.016487 -0.070519  0.747060  ...  0.101109   
1        ROCK  239200.0 -0.044436 -0.024889 -0.015285  ... -0.000394   
2        ROCK  170000.0 -0.044436 -0.024889 -0.015285  ... -0.000394   
3   LATINFOLK  237020.0 -0.044436 -0.024889 -0.015285  ... -0.000394   
4  FOREIGNRAP  298540.0 -0.044436 -0.024889 -0.015285  ... -0.000394   

     PCA_22    PCA_23    PCA_24    PCA_25    PCA_26    PCA_27    PCA_28  \
0 -0.018017  0.017332  0.020766  0.010944 -0.118673  0.024424  0.008445   
1 -0.000048  0.000101  0.000307  0.000150 -0.000701  0.000094  0.000337   
2 -0.000048  0.000101  0.000307  0.000150 -0.000701  0.000094  0.000337   
3 -0.000048  0.000101  0.000307  0.000150 -0.000701  0.000094  0.000337   
4 -0.000048  0.000101  0.000307  0.000150 -0.000701  0.000094  0.000337   

     PCA_29  days_since_release  
0  0.020194                1058  
1 -0.000018                 802  
2 -0.000018                4261  
3 -0.000018                 944  
4 -0.000018                3428  

[5 rows x 38 columns]

In [72]:
df_test.head()

track_id track_remake_type  text  title language  \
0  89803636c59346850bfb6de14370b303          ORIGINAL   211      7       RU   
1  224cb225a2e612f5489b24992a20cbf3             COVER     0     12       SV   
2  12f22354aa53f7a47bac64322f8a8779             COVER     0     22       EN   
3  0f61b0f76abab917e60b82b7fb58c6d6             COVER     0     17       PT   
4  c996348ac49e0c243a4eb4cb0a9cfbbb             COVER     0     20       PT   

      genres  duration     PCA_0    PCA_1     PCA_2  ...    PCA_21    PCA_22  \
0     RUSPOP  140990.0 -0.103916  0.73571  0.029124  ... -0.069664  0.048106   
1       FOLK  192310.0 -0.043729 -0.02442 -0.014640  ... -0.000121 -0.000004   
2      FILMS   64730.0 -0.043729 -0.02442 -0.014640  ... -0.000121 -0.000004   
3  LATINFOLK  138190.0 -0.043729 -0.02442 -0.014640  ... -0.000121 -0.000004   
4        POP  297330.0 -0.043729 -0.02442 -0.014640  ... -0.000121 -0.000004   

     PCA_23    PCA_24    PCA_25    PCA_26    PCA_27    PCA_28    PCA_29  \
0  0.037887 -0.012761  0.035715  0.029012  0.036190 -0.072929  0.249378   
1  0.000150 -0.000425 -0.000342 -0.000085 -0.000118 -0.000459 -0.000199   
2  0.000150 -0.000425 -0.000342 -0.000085 -0.000118 -0.000459 -0.000199   
3  0.000150 -0.000425 -0.000342 -0.000085 -0.000118 -0.000459 -0.000199   
4  0.000150 -0.000425 -0.000342 -0.000085 -0.000118 -0.000459 -0.000199   

   days_since_release  
0                 494  
1                 951  
2                1219  
3                 644  
4                 705  

[5 rows x 38 columns]

Проведем разделение датафрейма на тестовую и обущающие выборки. Размер тестовой выборки возьмем 25%

In [73]:
x_train = df_train.drop(['track_remake_type'], axis=1)
y_train = df_train[['track_remake_type']]
x_test = df_test.drop(['track_remake_type'], axis=1)
y_test = df_test[['track_remake_type']]

Заполним все пропуски заглушкой `Unknown`

In [74]:
x_train = x_train.fillna('Unknown')
x_test = x_test.fillna('Unknown')

Проведем кодирование категориальных признаков техникой `OrdinalEncoder`

In [75]:
encoder = OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1)

x_train_encoded = x_train.copy()
x_test_encoded = x_test.copy()

columns_to_encode = ['track_id' ,'language', 'genres']

x = pd.DataFrame(encoder.fit_transform(x_train_encoded[columns_to_encode]),
                               index=x_train_encoded.index, columns=columns_to_encode)

y = pd.DataFrame(encoder.transform(x_test_encoded[columns_to_encode]),
                              index=x_test_encoded.index, columns=columns_to_encode)

In [76]:
x_train_encoded.drop(columns_to_encode, axis=1, inplace=True)
x_test_encoded.drop(columns_to_encode, axis=1, inplace=True)

In [77]:
x_train_encoded = pd.merge(x_train_encoded, x, left_index=True, right_index=True)
x_test_encoded = pd.merge(x_test_encoded, y, left_index=True, right_index=True)

Целевой признак закодируем `LabelEncoder`

In [78]:
label_encoder = LabelEncoder()

y_train['track_remake_type'] = label_encoder.fit_transform(y_train['track_remake_type'])
y_test['track_remake_type'] = label_encoder.transform(y_test['track_remake_type'])

C:\Users\User\AppData\Local\Temp\ipykernel_16664\1775667603.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y_train['track_remake_type'] = label_encoder.fit_transform(y_train['track_remake_type'])
C:\Users\User\AppData\Local\Temp\ipykernel_16664\1775667603.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y_test['track_remake_type'] = label_encoder.transform(y_test['track_remake_type'])


Масштабируем численные признаки `StandartScaler`

In [79]:
numeric = ['text', 'title', 'duration', 'days_since_release']
scaler = StandardScaler()
scaler.fit(x_train[numeric])

x_train_scaled = pd.DataFrame(scaler.transform(x_train[numeric]))
x_test_scaled = pd.DataFrame(scaler.transform(x_test[numeric]))

x_train_scaled.index = x_train.index
x_test_scaled.index = x_test.index

In [80]:
x_train = pd.concat([x_train_scaled, x_train_encoded], axis=1)
x_test = pd.concat([x_test_scaled, x_test_encoded], axis=1)

In [81]:
x_train.columns = x_train.columns.astype(str)
x_test.columns = x_test.columns.astype(str)

In [82]:
y_train = np.ravel(y_train)
y_test = np.ravel(y_test)

In [83]:
x_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79805 entries, 0 to 79804
Data columns (total 41 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   0                   79805 non-null  float64
 1   1                   79805 non-null  float64
 2   2                   79805 non-null  float64
 3   3                   79805 non-null  float64
 4   text                79805 non-null  int64  
 5   title               79805 non-null  int64  
 6   duration            79805 non-null  float64
 7   PCA_0               79805 non-null  float64
 8   PCA_1               79805 non-null  float64
 9   PCA_2               79805 non-null  float64
 10  PCA_3               79805 non-null  float64
 11  PCA_4               79805 non-null  float64
 12  PCA_5               79805 non-null  float64
 13  PCA_6               79805 non-null  float64
 14  PCA_7               79805 non-null  float64
 15  PCA_8               79805 non-null  float64
 16  PCA_

In [84]:
x_train

0         1         2         3  text  title  duration  \
0      1.296232 -0.207092 -0.853886 -0.204730   157     14  132590.0   
1     -0.374019 -0.480441  0.404165 -0.455394     0     11  239200.0   
2     -0.374019 -0.207092 -0.412429  2.931511     0     14  170000.0   
3     -0.374019 -1.209370  0.378440 -0.316354     0      3  237020.0   
4     -0.374019 -0.024860  1.104407  2.115873     0     16  298540.0   
...         ...       ...       ...       ...   ...    ...       ...   
79800 -0.374019 -0.207092 -0.312597 -0.466165     0     14  178460.0   
79801 -0.374019 -0.389325  0.626014 -0.523935     0     12  258000.0   
79802 -0.374019 -0.571557  0.068559 -0.839224     0     10  210760.0   
79803  1.487726 -0.936022  0.204501  2.016978   175      6  222280.0   
79804 -0.374019 -0.571557  0.119891  0.790094     0     10  215110.0   

          PCA_0     PCA_1     PCA_2  ...    PCA_24    PCA_25    PCA_26  \
0     -0.016487 -0.070519  0.747060  ...  0.020766  0.010944 -0.118673   
1     -0.044436 -0.024889 -0.015285  ...  0.000307  0.000150 -0.000701   
2     -0.044436 -0.024889 -0.015285  ...  0.000307  0.000150 -0.000701   
3     -0.044436 -0.024889 -0.015285  ...  0.000307  0.000150 -0.000701   
4     -0.044436 -0.024889 -0.015285  ...  0.000307  0.000150 -0.000701   
...         ...       ...       ...  ...       ...       ...       ...   
79800 -0.044436 -0.024889 -0.015285  ...  0.000307  0.000150 -0.000701   
79801 -0.044436 -0.024889 -0.015285  ...  0.000307  0.000150 -0.000701   
79802 -0.044436 -0.024889 -0.015285  ...  0.000307  0.000150 -0.000701   
79803  0.430470  0.015648 -0.002655  ... -0.022725 -0.028391  0.003125   
79804 -0.044436 -0.024889 -0.015285  ...  0.000307  0.000150 -0.000701   

         PCA_27    PCA_28    PCA_29  days_since_release  track_id  language  \
0      0.024424  0.008445  0.020194                1058   16409.0      36.0   
1      0.000094  0.000337 -0.000018                 802   38289.0      16.0   
2      0.000094  0.000337 -0.000018                4261    5166.0      16.0   
3      0.000094  0.000337 -0.000018                 944   47116.0      17.0   
4      0.000094  0.000337 -0.000018                3428   52411.0      16.0   
...         ...       ...       ...                 ...       ...       ...   
79800  0.000094  0.000337 -0.000018                 791   19212.0      17.0   
79801  0.000094  0.000337 -0.000018                 732   58617.0      16.0   
79802  0.000094  0.000337 -0.000018                 410   11161.0      68.0   
79803  0.007227  0.017858  0.038015                3327   22676.0      59.0   
79804  0.000094  0.000337 -0.000018                2074   53896.0      16.0   

       genres  
0       107.0  
1        96.0  
2        96.0  
3        68.0  
4        52.0  
...       ...  
79800    68.0  
79801     1.0  
79802    29.0  
79803    26.0  
79804    61.0  

[79805 rows x 41 columns]

## Обучение моделей

### Baseline

In [85]:
dummy = DummyClassifier(strategy='uniform')

dummy.fit(x_train, y_train)
y_dummy = dummy.predict(x_test)
dummy_f1 = f1_score(y_test, y_dummy)
print(f'F1 константной модели: {dummy_f1}')

F1 константной модели: 0.1083678099559589


Дамми-модель выдала метрику 0.1, все остальные модели должны быть выше неё, чтобы быть адекватными.

### Решающее дерево

In [86]:
%%time
model = DecisionTreeClassifier(random_state=RANDOM_STATE, class_weight='balanced')
params = { 'max_depth': [15] }
grid_tree = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_tree.fit(x_train, y_train)
print(grid_tree.best_params_)
tree_score = grid_tree.best_score_
tree_score

{'max_depth': 15}
Wall time: 10.3 s


0.6961094403011894

У модели Решающего дерева метрика F1 существенно лучше

## Случайный лес

In [87]:
%%time
model = RandomForestClassifier(random_state=RANDOM_STATE, class_weight='balanced')
params = { 'n_estimators': [100] }
grid_forest = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_forest.fit(x_train, y_train)
print(grid_forest.best_params_)
random_forest_score = grid_forest.best_score_
random_forest_score

{'n_estimators': 100}
Wall time: 1min 17s


0.794595182267639

Модель случайного леса показывает существенно лучшую метрику F1

### Логистическая регрессия

In [88]:
%%time
model = LogisticRegression(random_state=RANDOM_STATE, class_weight='balanced')
params = { 'solver': ['liblinear'] }
grid_log = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_log.fit(x_train, y_train)
print(grid_log.best_params_)
logistic_regression_score = grid_log.best_score_
logistic_regression_score

{'solver': 'liblinear'}
Wall time: 11.3 s


0.39848843123842415

Сильно хуже

### LGBM

In [94]:
%%time
model = LGBMClassifier(random_state=RANDOM_STATE, class_weight='balanced')
params = { 'n_estimators': range(100, 500, 50),
           'max_depth': range(12, 24, 2),
           'learning_rate': [0.1]
         }
grid_lgb = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_lgb.fit(x_train, y_train)
print(grid_lgb.best_params_)
lgb_score = grid_lgb.best_score_
lgb_score = grid_lgb.best_score_
lgb_score

{'learning_rate': 0.1, 'max_depth': 16, 'n_estimators': 450}
Wall time: 2min 12s


0.8106112866277806

Модель LGBM показывает лучший результат

## AdaBoost

In [90]:
%%time
model = AdaBoostClassifier(random_state=RANDOM_STATE)
params = { 'n_estimators': [100],
           'learning_rate': [0.1] }
grid_ada = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_ada.fit(x_train, y_train)
print(grid_ada.best_params_)
adaboost_score = grid_ada.best_score_
adaboost_score

{'learning_rate': 0.1, 'n_estimators': 100}
Wall time: 53.8 s


0.6073678700296121

Результат плохой

### Тестирование

In [91]:
%%time
best_model = grid_lgb.best_estimator_
test_predictions = best_model.predict(x_test)
test_f1 = f1_score(y_test, test_predictions)
print(f'F1 на тестовой выборке: {test_f1}')

F1 на тестовой выборке: 0.7622764227642277
Wall time: 145 ms


Получили результат 0.81

## Отчет по проекту

В проекте необходимо было провести выбор модели машинного обучения и подбор ее гиперпараметров с целью классификации музыкального трека на оригинальный и кавер.

**Основные этапы выполнения проекта:**

*1. Исследовательский анализ данных.*
- провели ознакомление с данными;
- изменили тип данных;
- определили категориальные признаки;
- проверили данные на пропуски;
- заполнили пропуски;
- выявили влияние категориальных признаков на целевой;
- выявили корреляцию между признаками;
- ислледовали признаки на мультиколлинеарность.

*2. Подготовка данных для моделей:*
- объединили все таблицы в одну;
- добавили признаки `days_since_release` (число дней с даты релиза трека по сегодняшнее время)
- провели векторизацию текстов песен с целью добавления новых признаков;
- ввели новый признак - количество слов в названии трека
- провели преобразование категориальных признаков в численные техниками OrdinalEncoding;
- провели преобразование целевого признака техникой LabelEncoding;
- стандартизировали числовые признаки техникой StandartScaler;
- разделили данные на 2 выборки обучающую и тестовую в пропорции 3:1;
- проверили дисбаланс признаков;

*3. Обучение моделей.*
- Baseline;
- LogisticRegression;
- RandomForestClassifier;
- LightGBM;
- выявили лучшую модель по метрике F1

Перед выполнением проекта был составлен план. По итогу выполнения проекта план был выполнен в полном объеме.


В процессе выполнения проекта встретились с:

- в столбце `original_track_id` есть почти 400 дубликатов. Видимо это как раз те оригиналы, на которых есть больше чем 1 кавер.
- выяснили, что все треки у которых отсутствует`original_track_id` являются каверами. На основании этого заполнили пропуски.
- повстречали довольно большое количество пропусков, большую часть из них получилось заполнить. Остальное удалили.
- в поле `dttm` были довольно странные значения. Разобрались, преобразовали их в дату выхода трека.
- на основании анализа названий треков провели детекцию языка исполнения и заполнили пропуски в столбце `language`

Для обучения моделей, оценки параметров и проведения оценки важности признаков были написаны функции, позволяющие работать с любой моделью и любыми входными данными. Это добавило универсальности коду.

В результате проведенной работы были обучены 5 моделей. На обучающих выборках методом кросс-валидации были определены значения F1:

|Модель|F1|
|:-----|:--------|
|LogisticRegression|0.40|
|RandomForestClassifier|0.8|
|LGBMClassifier|0.81|
|AdaBoost|0.61|

На основании тестировании модели на тестовых данных итоговое значение метрики F1 составило **0.81**

Т.е. модель с точностью в 81% предсказывает является музыкальный трек оригиналом или кавером. Такая точность позволит быть достаточно уверенным что запрос пользователя будет адекватно обработан.

Для решения оставшихся задач по группировке каверов и нахождения их оригиналов, а также создания цепочки каверов с целью выхода на оригинал выявили нехватку данных. Есть предположение, что если заполнить все пропуски в поле `text` и провести NLP обработку текстов, то на основе частотного анализа возможно с большой точностью с взять треки между собой, используя KNN (метод ближайших соседей).

Далее, создать алгоритм, который на основе проведенной группировки и с учетом признака кавер/оригинал и даты релиза создать цепочки треков.

Основной вопрос - где взять тексты песен. Мы не смогли найти рессурс откуда возможно спарсить эти данные и при этом не нарушить законодательство в области авторского права. Возможно, самый простой способ - заказчику применить технологию распознования речи и преобразования её в текст.